### PR TITLE
Specialization for faster count(::Array{Bool})

### PR DESCRIPTION
**Pull request by jakobnissen**
See #34059. On my laptop, improves speed of `sum(::Array{Bool})` (and `count`) by about 8x. Added some tests to make sure it works for longer arrays (where the SIMD vectorization kicks in), and also works if the underlying data is not just `0x00` or `0x01`.

### CODE REVIEW

**Comment by recri** [comment]
Why is count(::Array{Bool}) masking Bool's with 1's?  Does BitArray(::Array{Bool}) mask Bool values?

**Comment by jakobnissen** [comment]
No, `BitArray{::Array{Bool}}` simply uses ordinary indexing into the boolean vector. However, indexing the boolean vector will implicitly insert a masking operation:
```
julia> f(x, i) = @inbounds x[i]
f (generic function with 2 methods)

julia> @code_native f([true, false], 1)
    .section    __TEXT,__text,regular,pure_instructions
; ┌ @ none:1 within `f'
; │┌ @ none:1 within `getindex'
    movq    (%rdi), %rax
    movb    -1(%rax,%rsi), %al
    andb    $1, %al
; │└
    retq
    nopw    (%rax,%rax)
; └
```
This is done to make sure that the binary representation of `true` is always `0x01` and `false` is always `0x00`. However, when the function is this PR uses `unsafe_load`, we have to do this masking operation ourselves.

**Comment by recri** [comment]
Okay, granted BitArray, even in the fast track using copy_to_bitarray_chunks! and pack8bools, is effectively masking to the low bit.  And I didn't think to check at_code_native to see how loads and stores get generated.

**Comment by haampie** [comment]
That's a nice speedup :)

> However, indexing the boolean vector will implicitly insert a masking operation. This is done to make sure that the binary representation of true is always 0x01 and false is always 0x00

At the end of the day this PR constructs the exact same masks, but at the level of 8 bytes (not a single byte). 

The speedup in this PR is probably not thanks to `popcnt` (i.e. `count_ones(...)`), but due to proper vectorization. The main difference is this PR generates code using `ymm` registers whereas the current master generates code using `xmm` registers. It might be worth figuring out why it does not vectorize well then.

**Comment by nalimilan** [comment]
Just a drive-by comment: it would be nice to add benchmarks to BaseBenchmarks to ensure this doesn't regress.

**Comment by jakobnissen** [comment]
@haampie 
If it's possible to make `count` perform better in a more generic, composable way by making the native code better, that would be a better solution than this specialization. However, I'm no good at understanding assembly level code, so I'm the wrong person to figure out why it doesn't vectorize well.

It's not due to the `xmm` versus `ymm` registrers, though. On my older CPU without `ymm`, both function uses `xmm`, yet this specialization is many times faster. Looking at the assembly for the old `count`, it looks like it computes the sum in a roundabout way using vectorized floating point operations and then extracts the bits. The specialized version doesn't actually use any bit counting instructions in the vectorized loop (on my older CPU) - it just creates better code. I'm not sure why.

File: base/reduce.jl [hunk 1]
```diff
@@ -861,3 +861,19 @@ function count(pred, a::AbstractArray)
     return n
 end
 count(itr) = count(identity, itr)
+
+function count(x::Array{Bool})
```
Comment (mbauman) [inline base/reduce.jl]:
```suggestion
function count(::typeof(identity), x::Array{Bool})
```

Let's override the more specific version here

**Comment by mbauman** [comment]
I agree with all of the above — this is worth having and it'd also be nice to make it a bit more generic.  It seems like loop unrolling can get us partway there (would have to add dispatching onto `IndexLinear` of course), but the solution here is still far and away better:

```julia
julia> using BenchmarkTools

julia> A = rand(Bool, 10_000);

julia> function count4(pred, a::AbstractArray)
           n = 0
           r = firstindex(a):4:lastindex(a)-3
           @inbounds @simd for i in r
               n += pred(a[i]) + pred(a[i+1]) + pred(a[i+2]) + pred(a[i+3])
           end
           @inbounds for i in last(r)+4:lastindex(a)
               n += pred(a[i])
           end
           return n
       end
count4 (generic function with 1 method)

julia> function count8(pred, a::AbstractArray)
           n = 0
           r = firstindex(a):8:lastindex(a)-7
           @inbounds @simd for i in r
               n += pred(a[i]) + pred(a[i+1]) + pred(a[i+2]) + pred(a[i+3]) +
                    pred(a[i+4]) + pred(a[i+5]) + pred(a[i+6]) + pred(a[i+7])
           end
           @inbounds for i in last(r)+8:lastindex(a)
               n += pred(a[i])
           end
           return n
       end
count8 (generic function with 1 method)

julia> @btime count4(identity, $A)
  658.450 ns (0 allocations: 0 bytes)
4948

julia> @btime count8(identity, $A)
  2.124 μs (0 allocations: 0 bytes)
4948

julia> @btime Base.count(identity, $A)
  1.751 μs (0 allocations: 0 bytes)
4948

julia> @btime count34060($A) # from 67ea908 
  223.839 ns (0 allocations: 0 bytes)
4948
```

File: base/reduce.jl [hunk 2]
```diff
@@ -861,3 +861,19 @@ function count(pred, a::AbstractArray)
...
 end
 count(itr) = count(identity, itr)
+
+function count(::typeof(identity), x::Array{Bool})
+    n = 0
+    chunks = length(x) ÷ sizeof(UInt)
+    mask = 0x0101010101010101 % UInt
+    GC.@preserve x begin
+        ptr = Ptr{UInt}(pointer(x))
+        for i in 1:chunks
```
Comment (stevengj) [inline base/reduce.jl]: Does `@simd` help?

Comment (KristofferC) [reply base/reduce.jl]: It shouldn't for integers, should it?

Comment (mbauman) [reply base/reduce.jl]: > It shouldn't for integers, should it?

I had the same thought, but it actually did help in my `count4` implementation above.  I think that's due to the different lowering of the iteration space for a step-range, though.

Given the timings and LLVM code for this implementation, I'd be very surprised if it did anything further.

**Comment by stev47** [comment]
I propose to represent chunked data as a data type for dispatch and reusability:
```julia
"""
    Chunk{T, I}

Represents data of length `sizeof(I)` as a chunk of elements of type `T`.
"""
struct Chunk{T, I}
    data::I
end
Chunk{T}(data::I) where {T,I} = Chunk{T,I}(data)

# type/operator specializations
Base.reduce(+, x::Chunk{Bool,UInt64}) = count_ones(x.data & 0x0101010101010101)

# generic implementation
function reduce_chunked(op, x::Array{T}) where T
    I = UInt
    ptr = Ptr{I}(pointer(x))
    chunks, tail = divrem(length(x), sizeof(I))
    acc = Base.reduce_empty(op, T)
    GC.@preserve x begin
        for i in 1:chunks
            y = Chunk{T}(unsafe_load(ptr, i))
            acc = op(acc, reduce(op, y))
        end
    end
    for i in length(x)-tail+1:length(x)
        acc = op(acc, x[i])
    end
    return acc
end

```
Performance is comparable to the pull-request.

If required, one could even generalize this concept further (simd-chunks vs word-chunks, etc.).

**Review by mbauman — APPROVED** [review]

**Pull request merged**